### DISCUSSION TITLE
Reasons for choosing the particular LMDB Java wrapper

### DISCUSSION THREAD

**Discussion by reckart** [eclipse-rdf4j/rdf4j#4853] [Q&A] [answered]
It appears that there are multiple LMDB Java wrappers. 

* https://github.com/lmdbjava/lmdbjava
* the one provided by LWJGL and which is currently used by RDF4J

Honestly, it feels strange to have to include a gaming library as a dependency when enabling the LMDB backend.

Did you consider the other wrapper? If yes, what was the reason for choosing the LWJGL wrapper?

**Comment by hmottestad** [comment]
@kenwenzel Any chance you have the time to take a look at this question?

**Comment by hmottestad** [comment]
One thing I noticed is that lmdbjava did not have aarch64 support for MacOS until now in December: https://github.com/lmdbjava/lmdbjava/pull/217 and https://github.com/lmdbjava/lmdbjava/wiki/Change-Log

I also noticed that lmdbjava is not yet in 1.0.0, which I assume to mean that the api is not yet stable.

**Comment by kenwenzel** [comment, marked as answer]
Hi @reckart,

I've tested lmdbjava but it was unstable and also has more overhead than LWJGL.
Although LWJGL has the word gaming in it's name, it is just a very good library for off-heap memory management and interoperability with native libraries. With LWJGL the LMDB functions can be almost used as if they were called from C or C++. The overhead of calling a function and working with pointers is extremely minimal. LWJGL is also planning for switching to Panama in the future.

Do you see any advantages of lmdbjava over the LWJGL bindings?

Best regards,
Ken

**Reply by reckart** [reply]
I've been using lmdbjava for a while since it was the first I found. I wasn't aware of the other. It is a single dependency instead of the like 2 deps per platform dependencies that lwjgl requires. That's a lazy reason of course since the split of platform support over multiple dependencies give better control over size vs. support.

So I don't know any hard advantages of lmdbjava vs. LWJGL. But thank you for your insight.

In fact, I consider giving the LWJGL version a try as it appears to be quite well maintained.

**Reply by kenwenzel** [reply]
I've had several problems with lmdbjava that were the main reason for dropping it. But also performance-wise you won't get any close to LWJGL's LMDB module with lmdbjava.
